### PR TITLE
Bug fix: Unresolved FK fields were not included in the FK's hash

### PR DESCRIPTION
**Pull request by fulghum**
A bug in `ForeignKey.HashOf()` was silently ignoring the `UnresolvedFKDetails` fields, due to using the wrong API to write the string values to the `bytes.Buffer`. Using `binary.Write` for a string (instead of `bytes.Buffer.Write`) was resulting in an error, but the error was not captured, so it was silently ignored. This led to FK name collisions when `@@FOREIGN_KEY_CHECKS` was disabled and the resolved column information wasn't populated to ensure a unique hash. 

Fixing this bug required several tests to be updated with new hash values. I also turned back on the unit tests in `foreign_key_test.go`, which had been skipped since the storage engine migration. I tried unsetting the `UnresolvedFKDetails` fields when the FK is resolved, but that broke many tests – it seems that the code still relies on the `UnresolvedFKDetails` fields being populated even when the FK is resolved. 

Related to: https://github.com/dolthub/dolt/issues/7384

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b8b29e6</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>b8b29e6</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b8b29e6</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/enginetest/dolt_transaction_queries.go [hunk 1]
```diff
@@ -1998,15 +1998,15 @@ var DoltConstraintViolationTransactionTests = []queries.TransactionTest{
 					"To allow transactions to be committed with constraint violations from a merge or transaction sequencing set @@dolt_force_transaction_commit=1.\n" +
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/enginetest/dolt_transaction_queries.go:1998]: I think we're missing a script test that exactly reproduces the original problem. Not completely clear on the exact setup, but something about setting `@foreign_key_checks` to 0 and creating two foreign keys that would have ended up with the same name before this change.

**Review by zachmu — APPROVED** [review]
LGTM, good catch!

Only significant comment is about a missing test.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8bc1433</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>8bc1433</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8bc1433</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/enginetest/dolt_transaction_queries.go (hunk 1 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_transaction_queries.go:1998]: Good call. I got too focused on cleaning up all those FK unit tests! I added a script test to trigger the same customer error; pretty easy to trigger once you turn off `@@foreign_key_checks`.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.998653 to 99.998653</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1aa48ff</TD>
<TD>not ok</TD>
<TD>80</TD>
</TR>
<TR><TD>1aa48ff</TD>
<TD>ok</TD>
<TD>5937377</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1aa48ff</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998653</TD>
</TR>
</table>

**Pull request merged**